src/Lean/Compiler/LCNF/PassManager.lean

```diff
@@ -11,27 +11,118 @@ import Lean.Compiler.LCNF.CompilerM
 
 namespace Lean.Compiler.LCNF
 
+/--
+The pipeline phase a certain `Pass` is supposed to happen in.
+-/
+inductive Phase where
+| /-- Here we still carry most of the original type information, most
+      of the dependent portion is already (partially) erased though. -/
+  base
+| /-- Here all types have been turned into `Any`. -/
+  untyped
+| /-- The lambda RC representation use for reference counting optimizations. -/
+  rc
```

Comment:
Yes, in the third phase we want to inject RC operations. We also want to eliminate the overhead of `BaseIO`. The general theme here is impurity. I think we should call this phase `impure`.